tools/prefetch_cache_gcsfuse/prefetch.go

```diff
@@ -0,0 +1,129 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/googlecloudplatform/gcsfuse/internal/gcsx"
+	"google.golang.org/api/iterator"
+)
+
+func downloadFile(ctx context.Context, client *storage.Client, object *storage.ObjectAttrs, cacheDir string) (err error) {
+	log.Printf(fmt.Sprintf("downloading file %v from bucket %v into dir %v", object.Name, object.Bucket, cacheDir))
+
+	// TODO ezl store the file name inside the json metadata instead of stripping the metadata filename
+	metadata := &gcsx.TempFileObjectMetadata{
+		BucketName:     object.Bucket,
+		ObjectName:     object.Name,
+		Generation:     object.Generation,
+		MetaGeneration: object.Metageneration,
+	}
+
+	// TODO ezl refactor once we decouple content cache from gcsx
+	// Create a temporary cache file on disk
+	// TODO ezl concerns about partial download and cache file corruption
+	// We may want a way to verify the files are fully downloaded
+	// and either resuming the download or discarding and redownloading the file
+	// We may also want to do cleanup if files are created on disk but aren't populated in time
+	f, err := ioutil.TempFile(cacheDir, gcsx.CACHE_FILE_PREFIX)
+
+	if err != nil {
+		err = fmt.Errorf("ioutil.TempFile: %w", err)
+		return
+	}
+	defer f.Close()
+
+	rc, err := client.Bucket(object.Bucket).Object(object.Name).NewReader(ctx)
+
+	if err != nil {
+		return fmt.Errorf("Object(%q).NewReader: %v", object.Name, err)
+	}
+	defer rc.Close()
+
+	if _, err := io.Copy(f, rc); err != nil {
+		return fmt.Errorf("io.Copy: %v", err)
+	}
+
+	file, err := json.MarshalIndent(*metadata, "", " ")
+	err = ioutil.WriteFile(fmt.Sprintf("%s.json", f.Name()), file, 0644)
+	if err != nil {
+		err = fmt.Errorf("downloadFile failed to write metadata: %w", err)
+	}
+	return
+}
+
+func prefetchCache(cacheDir, bucketName, prefix string) (err error) {
+	start := time.Now()
+	filesAttempted := 0
+	var filesDownloaded int32
+	var wg sync.WaitGroup
+
+	ctx := context.Background()
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		return fmt.Errorf("storage.NewClient: %w", err)
+	}
+	defer client.Close()
+
+	// Should we set a higher timeout or let this be configurable by the user
+	ctx, cancel := context.WithTimeout(ctx, time.Minute*5)
+	defer cancel()
+
+	it := client.Bucket(bucketName).Objects(ctx, &storage.Query{
+		Prefix: prefix,
+	})
+
+	// Concurrently download files from specified gcs bucket with prefix and delimiter
```

Comment:
This can potentially throttle the API or the disk by creating too many concurrent connections.

A better way of handling such a task is:
1. Create two channels: `pendingTasks` and `finishedTasks`
2. Create a pool of N workers by running `go func() {...}` for each worker as a goroutine. Then, each worker is going to take one `pendingTask`, do it, and put one in `finishedTask`.
3. The main goroutine calls `it.Next()` to get the objects, push them into the `pendingTasks`.
4. The main goroutine keeps getting the `finishedTasks` until all the tasks are finished.

When N = 1, this implementation becomes sequential.